astropy/table/table.py

```diff
@@ -278,8 +278,18 @@ def __init__(self, data=None, masked=None, names=None, dtype=None,
 
         elif data is None:
             if names is None:
-                return  # Empty table
-            else:
+                if dtype is None:
+                    return  # Empty table
+                try:
+                    # No data nor names but dtype is available.  This must be
+                    # valid to initialize a structured array.
+                    tmp = np.empty(0, dtype=dtype)
+                    names = tmp.dtype.names
+                    dtype = [tmp.dtype[i] for i in range(len(tmp.dtype))]
```

Comment:
Annoyingly / interestingly:
- `tmp.dtype` is not iterable
- `tmp.dtype[0]` is not the same as `tmp[names[0]].dtype`, so I could not iterate through `names`.